<center>
<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0101EN-Coursera/v2/M5_Final/images/SN_web_lightmode.png" width="300">
</center>


<h1>Analysis of Global COVID-19 Pandemic Data</h1>

Estimated time needed: **90** minutes



## Overview:

There are 10 tasks in this final project. All tasks will be graded by your peers who are also completing this assignment within the same session.

You need to submit the following the screenshot for the code and output for each task for review.

If you need to refresh your memories about specific coding details, you may refer to previous hands-on labs for code examples.


In [1]:
install.packages("rvest")

package 'rvest' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\Admin\AppData\Local\Temp\RtmpaIKVam\downloaded_packages


In [2]:
install.packages("httr")

package 'httr' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\Admin\AppData\Local\Temp\RtmpaIKVam\downloaded_packages


In [3]:
install.packages("stringr")

package 'stringr' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\Admin\AppData\Local\Temp\RtmpaIKVam\downloaded_packages


In [4]:
library(httr)
library(rvest)

In [5]:
library("stringr")

Note: if you can import above libraries, please use install.packages() to install them first.


## TASK 1: Get a `COVID-19 pandemic` Wiki page using HTTP request


First, let's write a function to use HTTP request to get a public COVID-19 Wiki page.

Before you write the function, you can open this public page from this 

URL https://en.wikipedia.org/w/index.php?title=Template:COVID-19_testing_by_country using a web browser.

The goal of task 1 is to get the html page using HTTP request (`httr` library)


## TASK 1: Get a `COVID-19 pandemic` Wiki page using HTTP request

In [6]:

get_wiki_covid19_page <- function() {
    
  # Our target COVID-19 wiki page URL is: https://en.wikipedia.org/w/index.php?title=Template:COVID-19_testing_by_country  
  # Which has two parts: 
    # 1) base URL `https://en.wikipedia.org/w/index.php  
    # 2) URL parameter: `title=Template:COVID-19_testing_by_country`, seperated by question mark ?
    

  # Wiki page base
    
  wiki_base_url <- "https://en.wikipedia.org/w/index.php"
    
  # You will need to create a List which has an element called `title` to specify which page you want to get from Wiki
  # in our case, it will be `Template:COVID-19_testing_by_country`
  
  url_query_param <- list(title = "Template:COVID-19_testing_by_country")   
    
  # - Use the `GET` function in httr library with a `url` argument and a `query` arugment to get a HTTP response
    #using two angled brackets in order to make "response" a global variable that can be reused outside the function'''
    response <<- GET(wiki_base_url, query=url_query_param)
    
  # Use the `return` function to return the response
#     response$status
    response$request$url
    response$status
}

Call the `get_wiki_covid19_page` function to get a http response with the target html page

In [7]:
# Call the get_wiki_covid19_page function and print the response
get_wiki_covid19_page()

[1] 200

## TASK 2: Extract COVID-19 testing data table from the wiki HTML page


On the COVID-19 testing wiki page, you should see a data table `<table>` node contains COVID-19 testing data by country on the page:

<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0101ENCoursera889-2022-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0101EN-Coursera/v2/M5_Final/images/covid-19-by-country.png" width="400" align="center">
</a>

Note the numbers you actually see on your page may be different from above because it is still an on-going pandemic when creating this notebook.

The goal of task 2 is to extract above data table and convert it into a data frame


Now use the `read_html` function in rvest library to get the root html node from response


In [8]:
# testing the content of the response variable
response

Response [https://en.wikipedia.org/w/index.php?title=Template%3ACOVID-19_testing_by_country]
  Date: 2022-12-01 06:40
  Status: 200
  Content-Type: text/html; charset=UTF-8
  Size: 417 kB
<!DOCTYPE html>
<html class="client-nojs" lang="en" dir="ltr">
<head>
<meta charset="UTF-8"/>
<title>Template:COVID-19 testing by country - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames...
"CS1 German-language sources (de)","CS1 Azerbaijani-language sources (az)","C...
"CS1 uses Japanese-language script (ja)","CS1 Japanese-language sources (ja)"...
"COVID-19 pandemic templates"],"wgPageContentLanguage":"en","wgPageContentMod...
"GEHomepageSuggestedEditsEnableTopics":true,"wgGETopicsMatchModeEnabled":fals...
...

In [9]:
content(response)

{html_document}
<html class="client-nojs" lang="en" dir="ltr">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8 ...
[2] <body class="skin-vector-legacy mediawiki ltr sitedir-ltr mw-hide-empty-e ...

In [10]:
response_headers <- headers(response)
response_headers

$date
[1] "Thu, 01 Dec 2022 06:40:56 GMT"

$server
[1] "mw1498.eqiad.wmnet"

$`x-content-type-options`
[1] "nosniff"

$`content-language`
[1] "en"

$vary
[1] "Accept-Encoding,Cookie,Authorization"

$`last-modified`
[1] "Wed, 30 Nov 2022 23:59:58 GMT"

$`content-type`
[1] "text/html; charset=UTF-8"

$`content-encoding`
[1] "gzip"

$age
[1] "0"

$`x-cache`
[1] "cp6016 miss, cp6009 miss"

$`x-cache-status`
[1] "miss"

$`server-timing`
[1] "cache;desc=\"miss\", host;desc=\"cp6009\""

$`strict-transport-security`
[1] "max-age=106384710; includeSubDomains; preload"

$`report-to`
[1] "{ \"group\": \"wm_nel\", \"max_age\": 86400, \"endpoints\": [{ \"url\": \"https://intake-logging.wikimedia.org/v1/events?stream=w3c.reportingapi.network_error&schema_uri=/w3c/reportingapi/network_error/1.0.0\" }] }"

$nel
[1] "{ \"report_to\": \"wm_nel\", \"max_age\": 86400, \"failure_fraction\": 0.05, \"success_fraction\": 0.0}"

$`set-cookie`
[1] "WMF-Last-Access=01-Dec-2022;Path=/;HttpOnly;secure;Expires=Mon, 02 Jan 2023 00:00:00 GMT"

$`set-cookie`
[1] "WMF-Last-Access-Global=01-Dec-2022;Path=/;Domain=.wikipedia.org;HttpOnly;secure;Expires=Mon, 02 Jan 2023 00:00:00 GMT"

$`accept-ch`
[1] "Sec-CH-UA-Arch,Sec-CH-UA-Bitness,Sec-CH-UA-Full-Version-List,Sec-CH-UA-Model,Sec-CH-UA-Platform-Version"

$`permissions-policy`
[1] "interest-cohort=(),ch-ua-arch=(self \"intake-analytics.wikimedia.org\"),ch-ua-bitness=(self \"intake-analytics.wikimedia.org\"),ch-ua-full-version-list=(self \"intake-analytics.wikimedia.org\"),ch-ua-model=(self \"intake-analytics.wikimedia.org\"),ch-ua-platform-version=(self \"intake-analytics.wikimedia.org\")"

$`x-client-ip`
[1] "154.68.225.162"

$`cache-control`
[1] "private, s-maxage=0, max-age=0, must-revalidate"

$`set-cookie`
[1] "GeoIP=NG:OS:Ile-Ife:7.48:4.56:v4; Path=/; secure; Domain=.wikipedia.org"

$`accept-ranges`
[1] "bytes"

$`transfer-encoding`
[1] "chunked"

attr(,"class")
[1] "insensitive" "list"

In [11]:
# Get the root html node from the http response in task 1
root_node <- read_html(response)

Get the correct table in the HTML root node using `html_node` function. 
Use the `inspect` function to check the live website and identify the attributes of the desired table.

In thi case, we're looking for the table with the following description: `<table class="wikitable plainrowheaders sortable collapsible autocollapse ...`

In [12]:
# Get the table nodes from the root html node
table_node <- html_nodes(root_node, "table")
table_node 

{xml_nodeset (3)}
[1] <table class="box-Update plainlinks ombox ombox-content ambox-Update" rol ...
[2] <table class="wikitable plainrowheaders sortable collapsible autocollapse ...
[3] <table class="plainlinks ombox mbox-small ombox-notice" role="presentatio ...

Scraping HTML tables with rvest: 
https://uc-r.github.io/scraping_HTML_tables

In [13]:
# getting the table node
second_tbl_node <- html_nodes(root_node, "table") %>% .[2]
second_tbl_node

{xml_nodeset (1)}
[1] <table class="wikitable plainrowheaders sortable collapsible autocollapse ...

Read the table node as a data frame using `html_table` function


## TASK 2: Extract COVID-19 testing data table from the wiki HTML page

In [14]:
raw_covid19_df <- html_table(table_node)
raw_covid19_df

[[1]]
# A tibble: 1 x 2
  X1    X2                                                                      
  <lgl> <chr>                                                                   
1 NA    This template needs to be updated.  Please help update this template to~

[[2]]
# A tibble: 173 x 9
   Country or reg~1 Date[~2 Tested Units~3 Confi~4 Confi~5 Teste~6 Confi~7 Ref. 
   <chr>            <chr>   <chr>  <chr>   <chr>   <chr>   <chr>   <chr>   <chr>
 1 Afghanistan      17 Dec~ 154,7~ samples 49,621  32.1    0.40    0.13    [1]  
 2 Albania          18 Feb~ 428,6~ samples 96,838  22.6    15.0    3.4     [2]  
 3 Algeria          2 Nov ~ 230,5~ samples 58,574  25.4    0.53    0.13    [3][~
 4 Andorra          23 Feb~ 300,3~ samples 37,958  12.6    387     49.0    [5]  
 5 Angola           2 Feb ~ 399,2~ samples 20,981  5.3     1.3     0.067   [6]  
 6 Antigua and Bar~ 6 Mar ~ 15,268 samples 832     5.4     15.9    0.86    [7]  
 7 Argentina        16 Apr~ 35,71~ samples 9,060,~ 25.4    78.3    20.0    [8]  
 8 Armenia          29 May~ 3,099~ samples 422,963 13.6    105     14.3    [9]  
 9 Australia        9 Sep ~ 78,54~ samples 10,112~ 12.9    313     40.3    [10] 
10 Austria          25 Nov~ 202,5~ samples 5,544,~ 2.7     2,275   62.3    [11] 
# ... with 163 more rows, and abbreviated variable names
#   1: `Country or region`, 2: `Date[a]`, 3: `Units[b]`, 4: `Confirmed(cases)`,
#   5: `Confirmed /tested,%`, 6: `Tested /population,%`,
#   7: `Confirmed /population,%`

[[3]]
# A tibble: 1 x 2
  X1    X2                                                                      
  <lgl> <chr>                                                                   
1 NA    "This template uses TemplateStyles:\nTemplate:COVID-19 testing by count~

In [15]:
# Read the table node and convert it into a data frame, and print the data frame for review
# Parsing the table nodes into a dataframe
covid_table <- html_table(second_tbl_node)
covid_table

[[1]]
# A tibble: 173 x 9
   Country or reg~1 Date[~2 Tested Units~3 Confi~4 Confi~5 Teste~6 Confi~7 Ref. 
   <chr>            <chr>   <chr>  <chr>   <chr>   <chr>   <chr>   <chr>   <chr>
 1 Afghanistan      17 Dec~ 154,7~ samples 49,621  32.1    0.40    0.13    [1]  
 2 Albania          18 Feb~ 428,6~ samples 96,838  22.6    15.0    3.4     [2]  
 3 Algeria          2 Nov ~ 230,5~ samples 58,574  25.4    0.53    0.13    [3][~
 4 Andorra          23 Feb~ 300,3~ samples 37,958  12.6    387     49.0    [5]  
 5 Angola           2 Feb ~ 399,2~ samples 20,981  5.3     1.3     0.067   [6]  
 6 Antigua and Bar~ 6 Mar ~ 15,268 samples 832     5.4     15.9    0.86    [7]  
 7 Argentina        16 Apr~ 35,71~ samples 9,060,~ 25.4    78.3    20.0    [8]  
 8 Armenia          29 May~ 3,099~ samples 422,963 13.6    105     14.3    [9]  
 9 Australia        9 Sep ~ 78,54~ samples 10,112~ 12.9    313     40.3    [10] 
10 Austria          25 Nov~ 202,5~ samples 5,544,~ 2.7     2,275   62.3    [11] 
# ... with 163 more rows, and abbreviated variable names
#   1: `Country or region`, 2: `Date[a]`, 3: `Units[b]`, 4: `Confirmed(cases)`,
#   5: `Confirmed /tested,%`, 6: `Tested /population,%`,
#   7: `Confirmed /population,%`

## TASK 3: Pre-process and export the extracted data frame

The goal of task 3 is to pre-process the extracted data frame from the previous step, and export it as a csv file


Let's get a summary of the data frame


In [16]:
# Print the summary of the data frame
summary(covid_table)

     Length Class  Mode
[1,] 9      tbl_df list

As you can see from the summary, the columns names are little bit different to understand and some column data types are not correct. For example, the `Tested` column shows as `character`. 

As such, the data frame read from HTML table will need some pre-processing such as removing irrelvant columns, renaming columns, and convert columns into proper data types.


We have prepared a pre-processing function for you to conver the data frame but you can also try to write one by yourself


In [17]:
preprocess_covid_data_frame <- function(data_frame) {
    
    shape <- dim(data_frame)

    # Remove the World row
    data_frame<-data_frame[!(data_frame$`Country or region`=="World"),]
    # Remove the last row
    data_frame <- data_frame[1:172, ]
    
    # We dont need the Units and Ref columns, so can be removed
    data_frame["Ref."] <- NULL
    data_frame["Units[b]"] <- NULL
    
    # Renaming the columns
    names(data_frame) <- c("country", "date", "tested", "confirmed", "confirmed.tested.ratio", "tested.population.ratio", "confirmed.population.ratio")
    
    # Convert column data types
    data_frame$country <- as.factor(data_frame$country)
    data_frame$date <- as.factor(data_frame$date)
    data_frame$tested <- as.numeric(gsub(",","",data_frame$tested))
    data_frame$confirmed <- as.numeric(gsub(",","",data_frame$confirmed))
    data_frame$'confirmed.tested.ratio' <- as.numeric(gsub(",","",data_frame$`confirmed.tested.ratio`))
    data_frame$'tested.population.ratio' <- as.numeric(gsub(",","",data_frame$`tested.population.ratio`))
    data_frame$'confirmed.population.ratio' <- as.numeric(gsub(",","",data_frame$`confirmed.population.ratio`))
    
    return(data_frame)
}

Call the `preprocess_covid_data_frame` function


In [18]:
# call `preprocess_covid_data_frame` function and assign it to a new data frame
covid19_df <- preprocess_covid_data_frame(covid_table)

ERROR: Error in data_frame[!(data_frame$`Country or region` == "World"), ]: incorrect number of dimensions


The `preprocess_covid_data_frame` probably failed because from the `summary` function, only one row was detected as against 173. As signified by [1,] in the description.

In [ ]:
covid19_df

Get the summary of the processed data frame again


In [ ]:
# Print the summary of the processed data frame again
summary(covid19_df)

After pre-processing, you can see the columns and columns names are simplified, and columns types are converted into correct types.


The data frame has following columns:

- **country** - The name of the country
- **date** - Reported date
- **tested** - Total tested cases by the reported date
- **confirmed** - Total confirmed cases by the reported date
- **confirmed.tested.ratio** - The ratio of confirmed cases to the tested cases
- **tested.population.ratio** - The ratio of tested cases to the population of the country
- **confirmed.population.ratio** - The ratio of confirmed cases to the population of the country


OK, we can call `write.csv()` function to save the csv file into a file. 


In [ ]:
# Export the data frame to a csv file
# file dir shortened afterwards
write.csv(covid19_df, file= "*/jupyternotebookfiles/coursera-R/final project/covid.csv", row.names=FALSE)

Note for IBM Waston Studio, there is no traditional "hard disk" associated with a R workspace.

Even if you call `write.csv()` method to save the data frame as a csv file, it won't be shown in IBM Cloud Object Storage asset UI automatically.

However, you may still check if the `covid.csv` exists using following code snippet:


In [ ]:
# Get working directory
#  wd <- getwd() # if we were saving to the working director
proj_folder <- '*/jupyternotebookfiles/coursera-R/final project' # shortened file dir


# Get exported 
#  file_path <- paste(wd, sep="", "/covid.csv") # if we saved to the working directory
file_path <- paste(proj_folder, sep="", "/covid.csv")

# File path
print(file_path)
file.exists(file_path)

**Optional Step**: If you have difficulties finishing above webscraping tasks, you may still continue with next tasks by downloading a provided csv file from here:


In [ ]:
# # Download a sample csv file
# covid_csv_file <- download.file("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0101EN-Coursera/v2/dataset/covid.csv", destfile="covid.csv")

## TASK 4: Get a subset of the extracted data frame

The goal of task 4 is to get the 5th to 10th rows from the data frame with only `country` and `confirmed` columns selected


We would be using the covid.csv file instead of the covid_new.csv file

In [22]:
# Read covid_data_frame_csv from the csv file
covid_data_frame_csv <- read.csv("covid.csv", header=TRUE, sep=",")
covid_data_frame_csv

country                date        tested    confirmed
1   Afghanistan            17 Dec 2020   154767     49621 
2   Albania                18 Feb 2021   428654     96838 
3   Algeria                2 Nov 2020    230553     58574 
4   Andorra                8 Mar 2021    159725     11066 
5   Angola                 12 Mar 2021   399228     20981 
6   Antigua and Barbuda    6 Mar 2021     15268       832 
7   Argentina              14 Mar 2021  7998673   2195722 
8   Armenia                13 Mar 2021   765879    177104 
9   Australia              15 Mar 2021 14933604     29130 
10  Austria                13 Mar 2021 17906847    488007 
11  Azerbaijan             14 Mar 2021  2710126    240295 
12  Bahamas                8 Mar 2021     70352      8642 
13  Bahrain                14 Mar 2021  3305543    131001 
14  Bangladesh             5 Mar 2021   4119031    549184 
15  Barbados               12 Mar 2021   133061      3404 
16  Belarus                13 Mar 2021  5063251    300146 
17  Belgium                12 Mar 2021 10110146    808283 
18  Belize                 12 Mar 2021    90633     12370 
19  Benin                  3 Mar 2021    504345      6071 
20  Bhutan                 14 Mar 2021   562368       868 
21  Bolivia                2 Mar 2021    731607    250557 
22  Bosnia and Herzegovina 12 Mar 2021   707190    142160 
23  Botswana               12 Mar 2021   856135     34108 
24  Brazil                 19 Feb 2021 23561497  10081676 
25  Brunei                 14 Mar 2021   110176       199 
26  Bulgaria               14 Mar 2021  1841263    278557 
27  Burkina Faso           4 Mar 2021    158777     12123 
28  Burundi                5 Jan 2021     90019       884 
29  Cambodia               2 Mar 2021    499716       844 
30  Cameroon               18 Feb 2021   942685     32681 
<U+22EE>   <U+22EE>               <U+22EE>    <U+22EE>  <U+22EE> 
143 Serbia                 14 Mar 2021   3149048   516277 
144 Singapore              8 Mar 2021    7805264    60052 
145 Slovakia               13 Mar 2021   2190451   336235 
146 Slovenia               14 Mar 2021    977071   200579 
147 South Africa           14 Mar 2021   9450492  1529420 
148 South Korea            1 Mar 2021    6592010    90029 
149 South Sudan            4 Mar 2021     113686     8527 
150 Spain                  11 Mar 2021  40292390  3178356 
151 Sri Lanka              12 Mar 2021   2230759    87286 
152 Sudan                  7 Jan 2021     158804    23316 
153 Sweden                 10 Mar 2021   6626099   693687 
154 Switzerland[l]         12 Mar 2021   5387481   570645 
155 Taiwan[m]              3 Mar 2021     429603      955 
156 Tanzania               18 Nov 2020      3880      509 
157 Thailand               4 Mar 2021    1579597    26162 
158 Togo                   5 Mar 2021     242657     7428 
159 Trinidad and Tobago    7 Mar 2021      98965     7736 
160 Tunisia                8 Mar 2021    1011915   238613 
161 Turkey                 14 Mar 2021  35126003  2879390 
162 Uganda                 11 Feb 2021    852444    39979 
163 Ukraine                13 Mar 2021   7328468  1451744 
164 United Arab Emirates   13 Mar 2021  33592518   424405 
165 United Kingdom         12 Mar 2021 103053938  4248286 
166 United States          14 Mar 2021 362125287 29403102 
167 Uruguay                14 Mar 2021   1140858    71691 
168 Uzbekistan             14 Jul 2020   1400000    13872 
169 Venezuela              11 Mar 2021   2987875   144277 
170 Vietnam                7 Mar 2021    2598753     2512 
171 Zambia                 3 Mar 2021    1106204    80090 
172 Zimbabwe               5 Mar 2021     392323    36248 
    confirmed.tested.ratio tested.population.ratio confirmed.population.ratio
1   32.10                    0.40                   0.1300                   
2   22.60                   15.00                   3.4000                   
3   25.40                    0.53                   0.1300                   
4    6.90                  

In [23]:
# Get the 5th to 10th rows, with two "country" "confirmed" columns
covid_data_frame_csv[5:10, c("country", "confirmed")]

country             confirmed
5  Angola                20981  
6  Antigua and Barbuda     832  
7  Argentina           2195722  
8  Armenia              177104  
9  Australia             29130  
10 Austria              488007

## TASK 5: Calculate worldwide COVID testing positive ratio

The goal of task 5 is to get the total confirmed and tested cases worldwide, and try to figure the overall positive ratio using `confirmed cases / tested cases`


In [24]:
# Get the total confirmed cases worldwide
sum(covid_data_frame_csv$confirmed) -> total_confirmed_cases_worldwide

# Get the total tested cases worldwide
sum(covid_data_frame_csv$tested) -> total_tested_cases_worldwide

# Get the positive ratio (confirmed / tested)
# ideally humans do not have decimals but this is a ratio so decimals are appropriate
positive_ratio <- as.numeric(total_confirmed_cases_worldwide/total_tested_cases_worldwide)
positive_ratio

[1] 0.06906584

## TASK 6: Get a country list which reported their testing data 

The goal of task 6 is to get a catalog or sorted list of countries who have reported their COVID-19 testing data


In [25]:
# Get the `country` column
country_tosort <- covid_data_frame_csv$country 
country_tosort

[1] "Afghanistan"            "Albania"                "Algeria"               
  [4] "Andorra"                "Angola"                 "Antigua and Barbuda"   
  [7] "Argentina"              "Armenia"                "Australia"             
 [10] "Austria"                "Azerbaijan"             "Bahamas"               
 [13] "Bahrain"                "Bangladesh"             "Barbados"              
 [16] "Belarus"                "Belgium"                "Belize"                
 [19] "Benin"                  "Bhutan"                 "Bolivia"               
 [22] "Bosnia and Herzegovina" "Botswana"               "Brazil"                
 [25] "Brunei"                 "Bulgaria"               "Burkina Faso"          
 [28] "Burundi"                "Cambodia"               "Cameroon"              
 [31] "Canada"                 "Chad"                   "Chile"                 
 [34] "China[c]"               "Colombia"               "Costa Rica"            
 [37] "Croatia"                "Cuba"                   "Cyprus[d]"             
 [40] "Czechia"                "Denmark[e]"             "Djibouti"              
 [43] "Dominica"               "Dominican Republic"     "DR Congo"              
 [46] "Ecuador"                "Egypt"                  "El Salvador"           
 [49] "Equatorial Guinea"      "Estonia"                "Eswatini"              
 [52] "Ethiopia"               "Faroe Islands"          "Fiji"                  
 [55] "Finland"                "France[f][g]"           "Gabon"                 
 [58] "Gambia"                 "Georgia[h]"             "Germany"               
 [61] "Ghana"                  "Greece"                 "Greenland"             
 [64] "Grenada"                "Guatemala"              "Guinea"                
 [67] "Guinea-Bissau"          "Guyana"                 "Haiti"                 
 [70] "Honduras"               "Hungary"                "Iceland"               
 [73] "India"                  "Indonesia"              "Iran"                  
 [76] "Iraq"                   "Ireland"                "Israel"                
 [79] "Italy"                  "Ivory Coast"            "Jamaica"               
 [82] "Japan"                  "Jordan"                 "Kazakhstan"            
 [85] "Kenya"                  "Kosovo"                 "Kuwait"                
 [88] "Kyrgyzstan"             "Laos"                   "Latvia"                
 [91] "Lebanon"                "Lesotho"                "Liberia"               
 [94] "Libya"                  "Lithuania"              "Luxembourg[i]"         
 [97] "Madagascar"             "Malawi"                 "Malaysia"              
[100] "Maldives"               "Mali"                   "Malta"                 
[103] "Mauritania"             "Mauritius"              "Mexico"                
[106] "Moldova[j]"             "Mongolia"               "Montenegro"            
[109] "Morocco"                "Mozambique"             "Myanmar"               
[112] "Namibia"                "Nepal"                  "Netherlands"           
[115] "New Caledonia"          "New Zealand"            "Niger"                 
[118] "Nigeria"                "North Korea"            "North Macedonia"       
[121] "Northern Cyprus[k]"     "Norway"                 "Oman"                  
[124] "Pakistan"               "Palestine"              "Panama"                
[127] "Papua New Guinea"       "Paraguay"               "Peru"                  
[130] "Philippines"            "Poland"                 "Portugal"              
[133] "Qatar"                  "Romania"                "Russia"                
[136] "Rwanda"                 "Saint Kitts and Nevis"  "Saint Lucia"           
[139] "Saint Vincent"          "San Marino"             "Saudi Arabia"          
[142] "Senegal"                "Serbia"                 "Singapore"             
[145] "Slovakia"               "Slovenia"               "South Africa"          
[148] "South Korea"            "S

In [26]:
# Check its class (should be Factor)
str(country_tosort)
class(country_tosort)

 chr [1:172] "Afghanistan" "Albania" "Algeria" "Andorra" "Angola" ...


[1] "character"

In [27]:
# Convert the country column into character so that you can easily sort them
as.character(country_tosort) -> country_tosort
country_tosort

[1] "Afghanistan"            "Albania"                "Algeria"               
  [4] "Andorra"                "Angola"                 "Antigua and Barbuda"   
  [7] "Argentina"              "Armenia"                "Australia"             
 [10] "Austria"                "Azerbaijan"             "Bahamas"               
 [13] "Bahrain"                "Bangladesh"             "Barbados"              
 [16] "Belarus"                "Belgium"                "Belize"                
 [19] "Benin"                  "Bhutan"                 "Bolivia"               
 [22] "Bosnia and Herzegovina" "Botswana"               "Brazil"                
 [25] "Brunei"                 "Bulgaria"               "Burkina Faso"          
 [28] "Burundi"                "Cambodia"               "Cameroon"              
 [31] "Canada"                 "Chad"                   "Chile"                 
 [34] "China[c]"               "Colombia"               "Costa Rica"            
 [37] "Croatia"                "Cuba"                   "Cyprus[d]"             
 [40] "Czechia"                "Denmark[e]"             "Djibouti"              
 [43] "Dominica"               "Dominican Republic"     "DR Congo"              
 [46] "Ecuador"                "Egypt"                  "El Salvador"           
 [49] "Equatorial Guinea"      "Estonia"                "Eswatini"              
 [52] "Ethiopia"               "Faroe Islands"          "Fiji"                  
 [55] "Finland"                "France[f][g]"           "Gabon"                 
 [58] "Gambia"                 "Georgia[h]"             "Germany"               
 [61] "Ghana"                  "Greece"                 "Greenland"             
 [64] "Grenada"                "Guatemala"              "Guinea"                
 [67] "Guinea-Bissau"          "Guyana"                 "Haiti"                 
 [70] "Honduras"               "Hungary"                "Iceland"               
 [73] "India"                  "Indonesia"              "Iran"                  
 [76] "Iraq"                   "Ireland"                "Israel"                
 [79] "Italy"                  "Ivory Coast"            "Jamaica"               
 [82] "Japan"                  "Jordan"                 "Kazakhstan"            
 [85] "Kenya"                  "Kosovo"                 "Kuwait"                
 [88] "Kyrgyzstan"             "Laos"                   "Latvia"                
 [91] "Lebanon"                "Lesotho"                "Liberia"               
 [94] "Libya"                  "Lithuania"              "Luxembourg[i]"         
 [97] "Madagascar"             "Malawi"                 "Malaysia"              
[100] "Maldives"               "Mali"                   "Malta"                 
[103] "Mauritania"             "Mauritius"              "Mexico"                
[106] "Moldova[j]"             "Mongolia"               "Montenegro"            
[109] "Morocco"                "Mozambique"             "Myanmar"               
[112] "Namibia"                "Nepal"                  "Netherlands"           
[115] "New Caledonia"          "New Zealand"            "Niger"                 
[118] "Nigeria"                "North Korea"            "North Macedonia"       
[121] "Northern Cyprus[k]"     "Norway"                 "Oman"                  
[124] "Pakistan"               "Palestine"              "Panama"                
[127] "Papua New Guinea"       "Paraguay"               "Peru"                  
[130] "Philippines"            "Poland"                 "Portugal"              
[133] "Qatar"                  "Romania"                "Russia"                
[136] "Rwanda"                 "Saint Kitts and Nevis"  "Saint Lucia"           
[139] "Saint Vincent"          "San Marino"             "Saudi Arabia"          
[142] "Senegal"                "Serbia"                 "Singapore"             
[145] "Slovakia"               "Slovenia"               "South Africa"          
[148] "South Korea"            "S

From the look of things, some country names contain unneccessary bits. For example, 'Georgia[h]', 'France[f][g]', 'Moldova[j]' and 'Northern Cyprus[k]' all have that extr bits in square brackets and that looks to be the trend throughout the column.

Match everything until Parenthesis: https://stackoverflow.com/questions/13867860/match-everything-until-parenthesis

In [28]:
# using regex to remove the extra bits on the country
as.list(sub(" *\\[.*", "", country_tosort)) -> country_aslist
country_aslist

[[1]]
[1] "Afghanistan"

[[2]]
[1] "Albania"

[[3]]
[1] "Algeria"

[[4]]
[1] "Andorra"

[[5]]
[1] "Angola"

[[6]]
[1] "Antigua and Barbuda"

[[7]]
[1] "Argentina"

[[8]]
[1] "Armenia"

[[9]]
[1] "Australia"

[[10]]
[1] "Austria"

[[11]]
[1] "Azerbaijan"

[[12]]
[1] "Bahamas"

[[13]]
[1] "Bahrain"

[[14]]
[1] "Bangladesh"

[[15]]
[1] "Barbados"

[[16]]
[1] "Belarus"

[[17]]
[1] "Belgium"

[[18]]
[1] "Belize"

[[19]]
[1] "Benin"

[[20]]
[1] "Bhutan"

[[21]]
[1] "Bolivia"

[[22]]
[1] "Bosnia and Herzegovina"

[[23]]
[1] "Botswana"

[[24]]
[1] "Brazil"

[[25]]
[1] "Brunei"

[[26]]
[1] "Bulgaria"

[[27]]
[1] "Burkina Faso"

[[28]]
[1] "Burundi"

[[29]]
[1] "Cambodia"

[[30]]
[1] "Cameroon"

[[31]]
[1] "Canada"

[[32]]
[1] "Chad"

[[33]]
[1] "Chile"

[[34]]
[1] "China"

[[35]]
[1] "Colombia"

[[36]]
[1] "Costa Rica"

[[37]]
[1] "Croatia"

[[38]]
[1] "Cuba"

[[39]]
[1] "Cyprus"

[[40]]
[1] "Czechia"

[[41]]
[1] "Denmark"

[[42]]
[1] "Djibouti"

[[43]]
[1] "Dominica"

[[44]]
[1] "Dominican Republic"

[[45]]
[1] "DR Congo"

[[46]]
[1] "Ecuador"

[[47]]
[1] "Egypt"

[[48]]
[1] "El Salvador"

[[49]]
[1] "Equatorial Guinea"

[[50]]
[1] "Estonia"

[[51]]
[1] "Eswatini"

[[52]]
[1] "Ethiopia"

[[53]]
[1] "Faroe Islands"

[[54]]
[1] "Fiji"

[[55]]
[1] "Finland"

[[56]]
[1] "France"

[[57]]
[1] "Gabon"

[[58]]
[1] "Gambia"

[[59]]
[1] "Georgia"

[[60]]
[1] "Germany"

[[61]]
[1] "Ghana"

[[62]]
[1] "Greece"

[[63]]
[1] "Greenland"

[[64]]
[1] "Grenada"

[[65]]
[1] "Guatemala"

[[66]]
[1] "Guinea"

[[67]]
[1] "Guinea-Bissau"

[[68]]
[1] "Guyana"

[[69]]
[1] "Haiti"

[[70]]
[1] "Honduras"

[[71]]
[1] "Hungary"

[[72]]
[1] "Iceland"

[[73]]
[1] "India"

[[74]]
[1] "Indonesia"

[[75]]
[1] "Iran"

[[76]]
[1] "Iraq"

[[77]]
[1] "Ireland"

[[78]]
[1] "Israel"

[[79]]
[1] "Italy"

[[80]]
[1] "Ivory Coast"

[[81]]
[1] "Jamaica"

[[82]]
[1] "Japan"

[[83]]
[1] "Jordan"

[[84]]
[1] "Kazakhstan"

[[85]]
[1] "Kenya"

[[86]]
[1] "Kosovo"

[[87]]
[1] "Kuwait"

[[88]]
[1] "Kyrgyzstan"

[[89]]
[1] "Laos"

[[90]]
[1] "Latvia"

[[91]]
[1] "Lebanon"

[[92]]
[1] "Lesotho"

[[93]]
[1] "Liberia"

[[94]]
[1] "Libya"

[[95]]
[1] "Lithuania"

[[96]]
[1] "Luxembourg"

[[97]]
[1] "Madagascar"

[[98]]
[1] "Malawi"

[[99]]
[1] "Malaysia"

[[100]]
[1] "Maldives"

[[101]]
[1] "Mali"

[[102]]
[1] "Malta"

[[103]]
[1] "Mauritania"

[[104]]
[1] "Mauritius"

[[105]]
[1] "Mexico"

[[106]]
[1] "Moldova"

[[107]]
[1] "Mongolia"

[[108]]
[1] "Montenegro"

[[109]]
[1] "Morocco"

[[110]]
[1] "Mozambique"

[[111]]
[1] "Myanmar"

[[112]]
[1] "Namibia"

[[113]]
[1] "Nepal"

[[114]]
[1] "Netherlands"

[[115]]
[1] "New Caledonia"

[[116]]
[1] "New Zealand"

[[117]]
[1] "Niger"

[[118]]
[1] "Nigeria"

[[119]]
[1] "North Korea"

[[120]]
[1] "North Macedonia"

[[121]]
[1] "Northern Cyprus"

[[122]]
[1] "Norway"

[[123]]
[1] "Oman"

[[124]]
[1] "Pakistan"

[[125]]
[1] "Palestine"

[[126]]
[1] "Panama"

[[127]]
[1] "Papua New Guinea"

[[128]]
[1] "Paraguay"

[[129]]
[1] "Peru"

[[130]]
[1] "Philippines"

[[131]]
[1] "Poland"

[[132]]
[1] "Portugal"

[[133]]
[1] "Qatar"

[[134]]
[1] "Romania"

[[135]]
[1] "Russia"

[[136]]
[1] "Rwanda"

[[137]]
[1] "Saint Kitts and Nevis"

[[138]]
[1] "Saint Lucia"

[[139]]
[1] "Saint Vincent"

[[140]]
[1] "San Marino"

[[141]]
[1] "Saudi Arabia"

[[142]]
[1] "Senegal"

[[143]]
[1] "Serbia"

[[144]]
[1] "Singapore"

[[145]]
[1] "Slovakia"

[[146]]
[1] "Slovenia"

[[147]]
[1] "South Africa"

[[148]]
[1] "South Korea"

[[149]]
[1] "South Sudan"

[[150]]
[1] "Spain"

[[151]]
[1] "Sri Lanka"

[[152]]
[1] "Sudan"

[[153]]
[1] "Sweden"

[[154]]
[1] "Switzerland"

[[155]]
[1] "Taiwan"

[[156]]
[1] "Tanzania"

[[157]]
[1] "Thailand"

[[158]]
[1] "Togo"

[[159]]
[1] "Trinidad and Tobago"

[[160]]
[1] "Tunisia"

[[161]]
[1] "Turkey"

[[162]]
[1] "Uganda"

[[163]]
[1] "Ukraine"

[[164]]
[1] "United Arab Emirates"

[[165]]
[1] "United Kingdom"

[[166]]
[1] "United States"

[[167]]
[1] "Uruguay"

[[168]]
[1] "Uzbekistan"

[[169]]
[1] "Venezuela"

[[170]]
[1] "Vietnam"

[[171

In [29]:
class(country_aslist)

[1] "list"

In [30]:
country_asvector <- unlist(country_aslist) 
country_asvector

[1] "Afghanistan"            "Albania"                "Algeria"               
  [4] "Andorra"                "Angola"                 "Antigua and Barbuda"   
  [7] "Argentina"              "Armenia"                "Australia"             
 [10] "Austria"                "Azerbaijan"             "Bahamas"               
 [13] "Bahrain"                "Bangladesh"             "Barbados"              
 [16] "Belarus"                "Belgium"                "Belize"                
 [19] "Benin"                  "Bhutan"                 "Bolivia"               
 [22] "Bosnia and Herzegovina" "Botswana"               "Brazil"                
 [25] "Brunei"                 "Bulgaria"               "Burkina Faso"          
 [28] "Burundi"                "Cambodia"               "Cameroon"              
 [31] "Canada"                 "Chad"                   "Chile"                 
 [34] "China"                  "Colombia"               "Costa Rica"            
 [37] "Croatia"                "Cuba"                   "Cyprus"                
 [40] "Czechia"                "Denmark"                "Djibouti"              
 [43] "Dominica"               "Dominican Republic"     "DR Congo"              
 [46] "Ecuador"                "Egypt"                  "El Salvador"           
 [49] "Equatorial Guinea"      "Estonia"                "Eswatini"              
 [52] "Ethiopia"               "Faroe Islands"          "Fiji"                  
 [55] "Finland"                "France"                 "Gabon"                 
 [58] "Gambia"                 "Georgia"                "Germany"               
 [61] "Ghana"                  "Greece"                 "Greenland"             
 [64] "Grenada"                "Guatemala"              "Guinea"                
 [67] "Guinea-Bissau"          "Guyana"                 "Haiti"                 
 [70] "Honduras"               "Hungary"                "Iceland"               
 [73] "India"                  "Indonesia"              "Iran"                  
 [76] "Iraq"                   "Ireland"                "Israel"                
 [79] "Italy"                  "Ivory Coast"            "Jamaica"               
 [82] "Japan"                  "Jordan"                 "Kazakhstan"            
 [85] "Kenya"                  "Kosovo"                 "Kuwait"                
 [88] "Kyrgyzstan"             "Laos"                   "Latvia"                
 [91] "Lebanon"                "Lesotho"                "Liberia"               
 [94] "Libya"                  "Lithuania"              "Luxembourg"            
 [97] "Madagascar"             "Malawi"                 "Malaysia"              
[100] "Maldives"               "Mali"                   "Malta"                 
[103] "Mauritania"             "Mauritius"              "Mexico"                
[106] "Moldova"                "Mongolia"               "Montenegro"            
[109] "Morocco"                "Mozambique"             "Myanmar"               
[112] "Namibia"                "Nepal"                  "Netherlands"           
[115] "New Caledonia"          "New Zealand"            "Niger"                 
[118] "Nigeria"                "North Korea"            "North Macedonia"       
[121] "Northern Cyprus"        "Norway"                 "Oman"                  
[124] "Pakistan"               "Palestine"              "Panama"                
[127] "Papua New Guinea"       "Paraguay"               "Peru"                  
[130] "Philippines"            "Poland"                 "Portugal"              
[133] "Qatar"                  "Romania"                "Russia"                
[136] "Rwanda"                 "Saint Kitts and Nevis"  "Saint Lucia"           
[139] "Saint Vincent"          "San Marino"             "Saudi Arabia"          
[142] "Senegal"                "Serbia"                 "Singapore"             
[145] "Slovakia"               "Slovenia"               "South Africa"          
[148] "South Korea"            "S

In [31]:
# Sort the countries atoz
country_asvector_atoz <- sort(country_asvector, decreasing = FALSE)
print(country_asvector_atoz)

  [1] "Afghanistan"            "Albania"                "Algeria"               
  [4] "Andorra"                "Angola"                 "Antigua and Barbuda"   
  [7] "Argentina"              "Armenia"                "Australia"             
 [10] "Austria"                "Azerbaijan"             "Bahamas"               
 [13] "Bahrain"                "Bangladesh"             "Barbados"              
 [16] "Belarus"                "Belgium"                "Belize"                
 [19] "Benin"                  "Bhutan"                 "Bolivia"               
 [22] "Bosnia and Herzegovina" "Botswana"               "Brazil"                
 [25] "Brunei"                 "Bulgaria"               "Burkina Faso"          
 [28] "Burundi"                "Cambodia"               "Cameroon"              
 [31] "Canada"                 "Chad"                   "Chile"                 
 [34] "China"                  "Colombia"               "Costa Rica"            
 [37] "Croatia"             

In [32]:
# Sort the countries ztoa
country_asvector_ztoa <- sort(country_asvector, decreasing = TRUE)
print(country_asvector_ztoa)

  [1] "Zimbabwe"               "Zambia"                 "Vietnam"               
  [4] "Venezuela"              "Uzbekistan"             "Uruguay"               
  [7] "United States"          "United Kingdom"         "United Arab Emirates"  
 [10] "Ukraine"                "Uganda"                 "Turkey"                
 [13] "Tunisia"                "Trinidad and Tobago"    "Togo"                  
 [16] "Thailand"               "Tanzania"               "Taiwan"                
 [19] "Switzerland"            "Sweden"                 "Sudan"                 
 [22] "Sri Lanka"              "Spain"                  "South Sudan"           
 [25] "South Korea"            "South Africa"           "Slovenia"              
 [28] "Slovakia"               "Singapore"              "Serbia"                
 [31] "Senegal"                "Saudi Arabia"           "San Marino"            
 [34] "Saint Vincent"          "Saint Lucia"            "Saint Kitts and Nevis" 
 [37] "Rwanda"              

In [33]:
# converting country vetor sorted in descending order back to list
country_list_ztoa <- list(country_asvector_ztoa)

# Print the sorted ZtoA list
print(country_list_ztoa)

[[1]]
  [1] "Zimbabwe"               "Zambia"                 "Vietnam"               
  [4] "Venezuela"              "Uzbekistan"             "Uruguay"               
  [7] "United States"          "United Kingdom"         "United Arab Emirates"  
 [10] "Ukraine"                "Uganda"                 "Turkey"                
 [13] "Tunisia"                "Trinidad and Tobago"    "Togo"                  
 [16] "Thailand"               "Tanzania"               "Taiwan"                
 [19] "Switzerland"            "Sweden"                 "Sudan"                 
 [22] "Sri Lanka"              "Spain"                  "South Sudan"           
 [25] "South Korea"            "South Africa"           "Slovenia"              
 [28] "Slovakia"               "Singapore"              "Serbia"                
 [31] "Senegal"                "Saudi Arabia"           "San Marino"            
 [34] "Saint Vincent"          "Saint Lucia"            "Saint Kitts and Nevis" 
 [37] "Rwanda"        

## TASK 7: Identify countries names with a specific pattern

The goal of task 7 is using a regular expression to find any countires start with `United`


Regular expression to match string starting with a specific word: 
https://stackoverflow.com/questions/1240504/regular-expression-to-match-string-starting-with-a-specific-word

In [34]:
# Use a regular expression `United.+` to find matches
pattern <- "United.+"

# Print the matched country names
print(str_extract(country_aslist, pattern))

  [1] NA                     NA                     NA                    
  [4] NA                     NA                     NA                    
  [7] NA                     NA                     NA                    
 [10] NA                     NA                     NA                    
 [13] NA                     NA                     NA                    
 [16] NA                     NA                     NA                    
 [19] NA                     NA                     NA                    
 [22] NA                     NA                     NA                    
 [25] NA                     NA                     NA                    
 [28] NA                     NA                     NA                    
 [31] NA                     NA                     NA                    
 [34] NA                     NA                     NA                    
 [37] NA                     NA                     NA                    
 [40] NA                 

## TASK 8: Pick two countries you are interested, and then review their testing data

The goal of task 8 is to compare the COVID-19 test data between two countires, you will need to select two rows from the dataframe, and select `country`, `confirmed`, `confirmed-population-ratio` columns


In [35]:
head(covid_data_frame_csv)

country             date        tested confirmed confirmed.tested.ratio
1 Afghanistan         17 Dec 2020 154767 49621     32.1                  
2 Albania             18 Feb 2021 428654 96838     22.6                  
3 Algeria             2 Nov 2020  230553 58574     25.4                  
4 Andorra             8 Mar 2021  159725 11066      6.9                  
5 Angola              12 Mar 2021 399228 20981      5.3                  
6 Antigua and Barbuda 6 Mar 2021   15268   832      5.4                  
  tested.population.ratio confirmed.population.ratio
1   0.40                   0.130                    
2  15.00                   3.400                    
3   0.53                   0.130                    
4 206.00                  14.300                    
5   1.30                   0.067                    
6  15.90                   0.860

In [36]:
# Select a subset (should be only one row) of data frame based on a selected country name and columns
nga_covid19 <- covid_data_frame_csv[118, c("country", "confirmed", "confirmed.tested.ratio")]
nga_covid19

# Select a subset (should be only one row) of data frame based on a selected country name and columns
uk_covid19 <- covid_data_frame_csv[165, c("country", "confirmed", "confirmed.tested.ratio")]
uk_covid19

country confirmed confirmed.tested.ratio
118 Nigeria 155657    10.1

country        confirmed confirmed.tested.ratio
165 United Kingdom 4248286   4.1

## TASK 9: Compare which one of the selected countries has a larger ratio of confirmed cases to population

The goal of task 9 is to find out which country you have selected before has larger ratio of confirmed cases to population, which may indicate that country has higher COVID-19 infection risk


In [37]:
# Using either nga_covid19$confirmed.tested.ratio OR as.numeric(nga_covid19[,3]) works in both cases
    # as they both return numeric datatypes when inspected with class() built-in method
        # the column name or data frame name can be changed to uk_covid19 as well

# I'll be sticking with 'dataframe$confirmed.tested.ratio' method  
# Because it is more descriptive and requires lesser functional nesting

# Use if-else statement
if (nga_covid19$confirmed.tested.ratio < uk_covid19$confirmed.tested.ratio) {
   print('There are more confirmed Covid-19 cases in Nigeria than in the United Kingdom')
} else if (nga_covid19$confirmed.tested.ratio == uk_covid19$confirmed.tested.ratio){
   print('There are as many confirmed Covid-19 cases in Nigeria as in the United Kingdom')
}  else {
   print('There are more confirmed Covid-19 cases in the United Kingdom than in Nigeria')
} 

[1] "There are more confirmed Covid-19 cases in the United Kingdom than in Nigeria"


## TASK 10: Find countries with confirmed to population ratio rate less than a threshold

The goal of task 10 is to find out which countries have the confirmed to population ratio less than 1%, it may indicate the risk of those countries are relatively low


How to print the column when another column met the condition:
https://stackoverflow.com/questions/41089883/how-to-print-the-column-when-another-column-met-the-condition

In [38]:
# Get a subset of any countries with `confirmed.population.ratio` less than the threshold
one_pct_threshold <- 1.0000

threshold_df <- covid_data_frame_csv[c("country", "confirmed.population.ratio")]
threshold_df

country                confirmed.population.ratio
1   Afghanistan             0.1300                   
2   Albania                 3.4000                   
3   Algeria                 0.1300                   
4   Andorra                14.3000                   
5   Angola                  0.0670                   
6   Antigua and Barbuda     0.8600                   
7   Argentina               4.8000                   
8   Armenia                 6.0000                   
9   Australia               0.1200                   
10  Austria                 5.5000                   
11  Azerbaijan              2.4000                   
12  Bahamas                 2.2000                   
13  Bahrain                 8.3000                   
14  Bangladesh              0.3300                   
15  Barbados                1.2000                   
16  Belarus                 3.2000                   
17  Belgium                 7.0000                   
18  Belize                  3.0000                   
19  Benin                   0.0520                   
20  Bhutan                  0.1200                   
21  Bolivia                 2.2000                   
22  Bosnia and Herzegovina  4.2000                   
23  Botswana                1.5000                   
24  Brazil                  4.8000                   
25  Brunei                  0.0430                   
26  Bulgaria                4.0000                   
27  Burkina Faso            0.0580                   
28  Burundi                 0.0074                   
29  Cambodia                0.0052                   
30  Cameroon                0.1200                   
<U+22EE>   <U+22EE>               <U+22EE>                  
143 Serbia                 7.40000                   
144 Singapore              1.10000                   
145 Slovakia               6.20000                   
146 Slovenia               9.60000                   
147 South Africa           2.60000                   
148 South Korea            0.17000                   
149 South Sudan            0.06700                   
150 Spain                  6.70000                   
151 Sri Lanka              0.40000                   
152 Sudan                  0.05300                   
153 Sweden                 6.70000                   
154 Switzerland[l]         6.60000                   
155 Taiwan[m]              0.00400                   
156 Tanzania               0.00085                   
157 Thailand               0.03800                   
158 Togo                   0.08600                   
159 Trinidad and Tobago    0.57000                   
160 Tunisia                2.00000                   
161 Turkey                 3.50000                   
162 Uganda                 0.08700                   
163 Ukraine                3.50000                   
164 United Arab Emirates   4.40000                   
165 United Kingdom         6.30000                   
166 United States          8.90000                   
167 Uruguay                2.10000                   
168 Uzbekistan             0.04100                   
169 Venezuela              0.50000                   
170 Vietnam                0.00250                   
171 Zambia                 0.46000                   
172 Zimbabwe               0.24000

In [39]:
# Again, we need to remove the unwanted parts of the country name
threshold_df$country <- sub(" *\\[.*", "", threshold_df$country)
threshold_df

country                confirmed.population.ratio
1   Afghanistan             0.1300                   
2   Albania                 3.4000                   
3   Algeria                 0.1300                   
4   Andorra                14.3000                   
5   Angola                  0.0670                   
6   Antigua and Barbuda     0.8600                   
7   Argentina               4.8000                   
8   Armenia                 6.0000                   
9   Australia               0.1200                   
10  Austria                 5.5000                   
11  Azerbaijan              2.4000                   
12  Bahamas                 2.2000                   
13  Bahrain                 8.3000                   
14  Bangladesh              0.3300                   
15  Barbados                1.2000                   
16  Belarus                 3.2000                   
17  Belgium                 7.0000                   
18  Belize                  3.0000                   
19  Benin                   0.0520                   
20  Bhutan                  0.1200                   
21  Bolivia                 2.2000                   
22  Bosnia and Herzegovina  4.2000                   
23  Botswana                1.5000                   
24  Brazil                  4.8000                   
25  Brunei                  0.0430                   
26  Bulgaria                4.0000                   
27  Burkina Faso            0.0580                   
28  Burundi                 0.0074                   
29  Cambodia                0.0052                   
30  Cameroon                0.1200                   
<U+22EE>   <U+22EE>               <U+22EE>                  
143 Serbia                 7.40000                   
144 Singapore              1.10000                   
145 Slovakia               6.20000                   
146 Slovenia               9.60000                   
147 South Africa           2.60000                   
148 South Korea            0.17000                   
149 South Sudan            0.06700                   
150 Spain                  6.70000                   
151 Sri Lanka              0.40000                   
152 Sudan                  0.05300                   
153 Sweden                 6.70000                   
154 Switzerland            6.60000                   
155 Taiwan                 0.00400                   
156 Tanzania               0.00085                   
157 Thailand               0.03800                   
158 Togo                   0.08600                   
159 Trinidad and Tobago    0.57000                   
160 Tunisia                2.00000                   
161 Turkey                 3.50000                   
162 Uganda                 0.08700                   
163 Ukraine                3.50000                   
164 United Arab Emirates   4.40000                   
165 United Kingdom         6.30000                   
166 United States          8.90000                   
167 Uruguay                2.10000                   
168 Uzbekistan             0.04100                   
169 Venezuela              0.50000                   
170 Vietnam                0.00250                   
171 Zambia                 0.46000                   
172 Zimbabwe               0.24000

In [40]:
class(threshold_df$country)

[1] "character"

In [41]:
threshold_df$country <- as.character(threshold_df$country)
class(threshold_df$country)

[1] "character"

In [42]:
# Print entry in the country column (threshold_df[,1]) 
# where entry in the confirmed.population.ratio(threshold_df[,2]) 
# is lesser than the 1% threshold
threshold_df[,1][threshold_df[,2] < one_pct_threshold]

[1] "Afghanistan"           "Algeria"               "Angola"               
 [4] "Antigua and Barbuda"   "Australia"             "Bangladesh"           
 [7] "Benin"                 "Bhutan"                "Brunei"               
[10] "Burkina Faso"          "Burundi"               "Cambodia"             
[13] "Cameroon"              "Chad"                  "China"                
[16] "Cuba"                  "Djibouti"              "Dominica"             
[19] "DR Congo"              "Egypt"                 "El Salvador"          
[22] "Equatorial Guinea"     "Ethiopia"              "Fiji"                 
[25] "Gabon"                 "Gambia"                "Ghana"                
[28] "Greenland"             "Grenada"               "Guinea"               
[31] "Guinea-Bissau"         "Haiti"                 "India"                
[34] "Indonesia"             "Ivory Coast"           "Jamaica"              
[37] "Japan"                 "Kenya"                 "Laos"                 
[40] "Lesotho"               "Liberia"               "Madagascar"           
[43] "Malawi"                "Malaysia"              "Mali"                 
[46] "Mauritania"            "Mauritius"             "Mongolia"             
[49] "Montenegro"            "Mozambique"            "Myanmar"              
[52] "Nepal"                 "New Caledonia"         "New Zealand"          
[55] "Niger"                 "Nigeria"               "North Korea"          
[58] "Pakistan"              "Papua New Guinea"      "Philippines"          
[61] "Rwanda"                "Saint Kitts and Nevis" "Senegal"              
[64] "South Korea"           "South Sudan"           "Sri Lanka"            
[67] "Sudan"                 "Taiwan"                "Tanzania"             
[70] "Thailand"              "Togo"                  "Trinidad and Tobago"  
[73] "Uganda"                "Uzbekistan"            "Venezuela"            
[76] "Vietnam"               "Zambia"                "Zimbabwe"